remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -422,75 +428,134 @@ impl REClient {
 
     pub async fn upload(
         &self,
-        metadata: RemoteExecutionMetadata,
+        _metadata: RemoteExecutionMetadata,
         request: UploadRequest,
     ) -> anyhow::Result<UploadResponse> {
-        let mut client = self.grpc_clients.cas_client.clone();
-
-        let files_with_digest: Vec<Request> = request
-            .files_with_digest
-            .unwrap_or_default()
-            .into_try_map(|x| {
-                anyhow::Ok(Request {
-                    digest: Some(tdigest_to(x.digest)),
-                    // FIXME: This could do a lot of blocking reads
-                    data: fs_util::read(&x.name)?,
-                    compressor: compressor::Value::Identity as i32,
-                })
-            })?;
-
-        let re_request = BatchUpdateBlobsRequest {
-            instance_name: INSTANCE_NAME.into(),
-            requests: [
-                request
-                    .inlined_blobs_with_digest
-                    .unwrap_or_default()
-                    .into_map(|x| Request {
-                        digest: Some(tdigest_to(x.digest)),
-                        data: x.blob,
-                        compressor: compressor::Value::Identity as i32,
-                    }),
-                files_with_digest,
-            ]
-            .concat(),
+        let mut bs_client = self.grpc_clients.bytestream_client.clone();
+
+        /* 4MB, with some slop just in case */
+        /* XXX FIXME (aseipp): can this come from grpc transport? CacheCapabilities? */
+        let grpc_msg_limit = (4 * 1000 * 1000) - 4096;
+
+        let mut blob_hashes = vec![];
+
+        let instance_prefix = if let Some(name) = self.instance_name.clone() {
+            format!("{}/", name)
+        } else {
+            "".to_string()
         };
 
-        let blob_hashes = re_request
-            .requests
-            .iter()
-            .map(|x| x.digest.as_ref().unwrap().hash.clone())
-            .collect::<Vec<String>>();
-        let response = client
-            .batch_update_blobs(with_internal_metadata(re_request, metadata))
-            .await?;
+        // upload inlined blobs
+        let inlined_blobs = request.inlined_blobs_with_digest;
+        for blob in inlined_blobs.unwrap_or_default() {
+            let hash = blob.digest.hash.clone();
+            let size = blob.digest.size_in_bytes;
+            let uuid = uuid::Uuid::new_v4().to_string();
+            let data = blob.blob;
+
+            let resource_name = format!(
+                "{}uploads/{}/blobs/{}/{}",
+                instance_prefix, uuid, hash, size
+            );
 
```

Comment:
Can we point to the underlying docs for this? And ideally I think you might be supposed to actually send the digest type here IIRC (though not in all cases if I read the spec right)?

Reply:
You must not specify digest_function here if you're not using one of MD5, MURMUR3, SHA1, SHA256, SHA384, SHA512, or VSO hashes.

If you're using SHA256TREE or BLAKE3 (and probably any hashes added to the spec from now on), then you must specify digest_function here.

https://github.com/bazelbuild/remote-apis/blob/35aee1c4a4250d3df846f7ba3e4a4e66cb014ecd/build/bazel/remote/execution/v2/remote_execution.proto#L227